README.md
# cljnewbiebank

Intended to tackle on Clojure's idiomatic DB and REST, 
this repo was created for study matters, focusing on retrieving simple 
information from database and REST endpoints. 
It is not of my intention to deliver an awesome, full-powered and "production 
deployable" app.


## Models
The purpose of this section is to describe the tables that were used on the schema.
The data is maintained on postgresql-heroku.

They are very simple and straightforward. In a real app the tables should have
more fields: the `users` could hold more informations about the user 
(country, city, zip code), etc.
However, I just created the schema to have a structure in which I could test the
DB approaches and write some Clojure. 

### users
>iduser, name, createdat

### accounts
>idaccount, iduser, createdat, balance

## transactions_types
>cdtransactiontype, description

### transactions
>idtransaction, createdat, amount, acc_to, acc_from

An user can hold 1 (or more) accounts, those who can trade money by transactions.
Each transaction holds the amount, a holder for the type of the transaction 
and the accounts.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-06-02
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-06-02
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/cljnewbiebank/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/cljnewbiebank/compare/0.1.0...0.1.1


test/cljnewbiebank/core_test.clj
(ns cljnewbiebank.core-test
  (:require [clojure.test :refer :all]
            [cljnewbiebank.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 1 1))))


test/cljnewbiebank/database_test.clj
(ns cljnewbiebank.database-test
  (:require [clojure.test :refer :all]
            [cljnewbiebank.database :as db]))

; users table
(def test-user {:createdat (db/string-to-timestamp "2018-06-02 20:13:00") 
                :name "teste lein" 
                :email "foo@bar.foo"})
(def insert-user #(db/create-user %))

(defn create-user-test []
    (not= nil (insert-user test-user)))

(defn update-user-test []
    (let [dummy-user (first (db/get-all-users))
          dummy-id (:iduser dummy-user)]
        (= '(1) (db/update-user dummy-id {:name "database_test.update-user-test"}))
    )
)

(defn delete-user-test []
    (let [dummy-user (first (db/get-all-users))
          dummy-id (:iduser dummy-user)]
      (= '(1) (db/delete-user dummy-id))))
  

; accounts table
(def test-account {:iduser (:iduser (first (db/get-all-users)))
                   :createdat (db/string-to-timestamp "2018-06-02 20:13:00") 
                   :balance 500.00})
(def insert-account #(db/create-account %))

(defn create-account-test []
    (not= nil (insert-account test-account)))

(defn update-account-test []
    (let [dummy-account (first (db/get-all-accounts))
          dummy-id (:idaccount dummy-account)]
        (= '(1) (db/update-account dummy-id {:balance 450.0}))
    )
)

(defn delete-account-test []
    (let [dummy-account (first (db/get-all-accounts))
          dummy-id (:idaccount dummy-account)]
      (= '(1) (db/delete-account dummy-id))))
  
;transactions
(def test-transaction {:createdat (db/string-to-timestamp "2018-06-02 20:13:00") 
                       :cdtransactiontype 1 ; deposit
                       :amount 100
                       :acc_to (:idaccount (first (db/get-all-accounts)))})
(def insert-transaction #(db/create-transaction %))

(defn create-transaction-test []
    (not= nil (insert-transaction test-transaction)))

(deftest database-test
    (testing "INSERT on users" (is (create-user-test)))
    (testing "DELETE on users" (is (delete-user-test)))
    (testing "INSERT on users" (is (create-user-test)))
    (testing "UPDATE on users" (is (update-user-test)))
    (testing "INSERT on accounts" (is (create-account-test)))
    (testing "DELETE on accounts" (is (delete-account-test)))
    (testing "INSERT on accounts" (is (create-account-test)))
    (testing "UPDATE on accounts" (is (update-account-test)))   
    (testing "INSERT on transactions" (is (create-transaction-test)))
)
    
; lein test cljnewbiebank.database-test


project.clj
(defproject cljnewbiebank "0.1.0-SNAPSHOT"
  :description "The wonderful bank of Clojure's newbieland"
  :url "https://github.com/vitorhead/clojure-newbie"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [org.clojure/java.jdbc "0.3.7"]
                 [org.postgresql/postgresql "9.4.1212"]
                 [clj-time "0.13.0"]
                 [compojure "1.1.1"]
                 [ring/ring-json "0.1.2"]]
  ;:main ^:skip-aot cljnewbiebank.core
  :plugins [[lein-ring "0.7.3"]]
  :ring {:handler cljnewbiebank.handler/app
         :auto-reload? true
         :auto-refresh? false
         :open-browser? false}    
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}})


src/cljnewbiebank/database.clj
(ns cljnewbiebank.database
    (:require [clojure.java.jdbc :as sql]
              [clojure.string :as str]
              [clj-time.coerce :as timec]
              [clj-time.format :as timef]
              [clj-time.core :as timecore]))
          

; Timestamp conversions
; These function can be used to convert a string into a timestamp and then insert it on the db
(defn string-to-timestamp
    [st]
    (timec/to-sql-date st))

(defn timestamp-to-string
    [ts]
    (timef/unparse (timef/formatters :date) (timec/from-sql-date ts)))

; DB Connection
; postgres://[user]:[password]@[host]:[port]/[database]
(def postgresql {:dbtype ""
                 :dbname ""
                 :host ""
                 :port ""
                 :user ""
                 :password ""
                 :ssl true
                 :sslfactory ""})                                  

; ==               
; Users table
; ==

; To convert the timestamp instance
(def user-from-db {:iduser identity,
                   :name identity, 
                   :createdat timestamp-to-string})

(defn create-user 
    [user]
    (let [parsed-user (into user 
                            {:createdat (timec/to-timestamp (timecore/now))}
                            )]
        (sql/insert! postgresql
                     :users parsed-user)))
           
(defn get-all-users []
    (into [] 
        (sql/query postgresql 
            ["SELECT * FROM users ORDER BY idUser DESC"])))
        
(defn get-user 
    [id]
    (sql/query postgresql 
              ["SELECT * FROM users WHERE idUser = ?" id])
)

(defn update-user
    [id user]
    (sql/update! postgresql :users user ["idUser = ?" id]))


(defn delete-user
    [id]
    (sql/delete! postgresql :users ["idUser = ?" id]))


; ==
; Accounts table
; ==

(defn create-account 
    [account]
    (let [parsed-account (into account 
                            {:createdat (timec/to-timestamp (timecore/now))}
                            )]    
        (sql/insert! postgresql
                     :accounts parsed-account)))
           
(defn get-all-accounts []
    (into [] 
        (sql/query postgresql 
            ["SELECT * FROM accounts ORDER BY idAccount DESC"])))
        
(defn get-account 
    [id]
    (sql/query postgresql 
              ["SELECT * FROM accounts WHERE idAccount = ?" id])
)

(defn update-account
    [id account]
    (sql/update! postgresql :accounts account ["idAccount = ?" id]))


(defn delete-account
    [id]
    (sql/delete! postgresql :accounts ["idAccount = ?" id]))

; ==
; Transactions table
; ==

(defn create-transaction 
    [transaction]
    (let [parsed-transaction (into transaction 
                              {:createdat (timec/to-timestamp (timecore/now))}
                             )]      
      (sql/insert! postgresql
                   :transactions parsed-transaction)))
           
(defn get-all-transactions []
    (into [] 
        (sql/query postgresql 
            ["SELECT * FROM transactions ORDER BY idTransaction DESC"])))
        
(defn get-transaction 
    [id]
    (sql/query postgresql 
              ["SELECT * FROM transactions WHERE idTransaction = ?" id])
)


src/cljnewbiebank/context/accounts.clj
(ns cljnewbiebank.context.accounts
    (:use ring.util.response)
    (:require [cljnewbiebank.database :as db]
              [clj-time.coerce :as timec]
              [clj-time.format :as timef]
              [clj-time.core :as timecore]
              [clojure.java.jdbc :as sql]))
          
; ==
; Accounts table
; ==

(defn insert-account 
    [account]
    (let [parsed-account (into account 
                            {:createdat (timec/to-timestamp (timecore/now))}
                            )]    
        (sql/insert! db/postgresql
                     :accounts parsed-account)))
           
(defn select-all-accounts []
    (into [] 
        (sql/query db/postgresql 
            ["SELECT * FROM accounts ORDER BY idAccount DESC"])))
        
(defn select-account 
    [id]
    (sql/query db/postgresql 
              ["SELECT * FROM accounts WHERE idAccount = ?" id])
)

(defn update-account
    [id account]
    (sql/update! db/postgresql :accounts account ["idAccount = ?" id]))


(defn delete-account
    [id]
    (sql/delete! db/postgresql :accounts ["idAccount = ?" id]))

          
          
; == 
; Accounts endpoint
; ==
(defn get-accounts []
    (cljnewbiebank-response (select-all-accounts)))

(defn get-account 
    [id]
    (cljnewbiebank-response (db/get-account (read-string id))))


; curl -d '{"iduser": 35, "balance": 50.0}' -H "Content-Type: application/json" -X POST localhost:8080/accounts
(defn post-account
    [body]
    (db/create-account body)
    {:status 201})

(defn put-account
    [id body]
    (db/update-account id body)
    (get-account id) ;return updated account
)

(defn delete-account
    [id]
    (db/delete-account id)
    {:status 204})

src/cljnewbiebank/handler.clj
(ns cljnewbiebank.handler
    (:use compojure.core)
    (:use ring.util.response)
    (:require [compojure.handler :as handler]
              [ring.middleware.json :as middleware]
              [compojure.route :as route]
              [cljnewbiebank.database :as db]))

(defn cljnewbiebank-response 
    "This function returns the content wrapped in a result tag"
    [content]
    (response {:result content}))

; ==
; Accounts endpoints
; ==

(defn get-accounts []
    (cljnewbiebank-response (db/get-all-accounts)))

(defn get-account 
    [id]
    (cljnewbiebank-response (db/get-account (read-string id))))


; curl -d '{"iduser": 35, "balance": 50.0}' -H "Content-Type: application/json" -X POST localhost:8080/accounts
(defn post-account
    [body]
    (db/create-account body)
    {:status 201})

(defn put-account
    [id body]
    (db/update-account id body)
    (get-account id) ;return updated account
)

(defn delete-account
    [id]
    (db/delete-account id)
    {:status 204})

; ==
; Users endpoints
; ==

(defn get-users []
    (cljnewbiebank-response (db/get-all-users)))

(defn get-user 
    [id]
    (cljnewbiebank-response (db/get-user (read-string id))))


; curl -d '{"name": "endpoint test", "email": "endpoint@endpoint.com"}' -H "Content-Type: application/json" -X POST localhost:8080/users
(defn post-user
    [body]
    (db/create-user body)
    {:status 201})

(defn put-user
    [id body]
    (db/update-user id body)
    (get-user id) ;return updated user
)

(defn delete-user
    [id]
    (db/delete-user id)
    {:status 204})

; == 
; Transactions endpoints
; ==
(defn get-transactions []
    (cljnewbiebank-response (db/get-all-transactions)))

(defn get-transaction 
    [id]
    (cljnewbiebank-response (db/get-transaction (read-string id))))


; curl -d '{"cdtransactiontype": 3, "amount": 25.0, "acc_to": 8, "acc_from": 11}' -H "Content-Type: application/json" -X POST localhost:8080/transactions
(defn post-transaction
    [body]
    (db/create-transaction body)
    {:status 201})

(defroutes cljnewbiebank-routes
  (context "/accounts" [] (defroutes accounts-routes
    (GET "/" [] (get-accounts))
    (POST "/" {body :body} (post-account body))
    (context "/:id" [id] (defroutes account-routes 
    ; :id is a placeholder and can then be injected into our parameter vector
      (GET "/" [] (get-account id)) 
      (PUT "/" {body :body} (put-account id body))
      (DELETE "/" [] (delete-account id))
    ))))
    
  (context "/users" [] (defroutes users-routes
    (GET "/" [] (get-users))
    (POST "/" {body :body} (post-user body))
    (context "/:id" [id] (defroutes user-routes
      (GET "/" [] (get-user id))
      (PUT "/" {body :body} (put-user id body))
      (DELETE "/" [] (delete-user id))
      ))))
  
  (context "/transactions" [] (defroutes transactions-routes
    (GET "/" [] (get-transactions))
    (POST "/" {body :body} (post-transaction body))
    (context "/:id" [id] (defroutes transaction-routes
      (GET "/" [] (get-transaction id))
      ))))
  
  (route/not-found "not found :(")
)
          
(def app
    (-> (handler/api cljnewbiebank-routes)
        (middleware/wrap-json-body) ;; parse json bodies
        (middleware/wrap-json-response))) ;; wrap responses in the ws